beater/middleware/timeout_middleware.go

```diff
@@ -0,0 +1,45 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package middleware
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/apm-server/beater/request"
+)
+
+// TimeoutMiddleware assumes that a context.Canceled error indicates a timed out
+// request. This could be caused by a either a client timeout or server timeout.
+// The middleware sets the Context.Result.
+func TimeoutMiddleware() Middleware {
+	tErr := errors.New("request timed out")
+	return func(h request.Handler) (request.Handler, error) {
+		return func(c *request.Context) {
+			h(c)
+
+			err := c.Request.Context().Err()
+			if errors.Is(err, context.Canceled) {
+				c.Result.SetDefault(request.IDResponseErrorsTimeout)
+				c.Result.Err = tErr
+				c.Result.Body = tErr.Error()
```

Comment:
IIANM, Result.SetDefault indirectly calls Result.Set, which sets these:

https://github.com/elastic/apm-server/blob/09cbdea568f8e43b91b197b4fc81e8662c88e9ce/beater/request/result.go#L194-L195

Was something failing that was fixed by making this change?

Reply:
Kinda sorta. The body wasn't being set, and the error was being wrapped. My initial error was no longer being used so compilation failed, and then I read the code further and saw that it wasn't setting the body, and then I opted to set the error explicitly because part of the github issue was removing "context canceled" since it doesn't carry meaning for non-go devs.

https://github.com/elastic/apm-server/blob/09cbdea568f8e43b91b197b4fc81e8662c88e9ce/beater/request/result.go#L160-L162

https://github.com/elastic/apm-server/blob/09cbdea568f8e43b91b197b4fc81e8662c88e9ce/beater/request/result.go#L216-L222

We should probably set the body, otherwise it's `nil`, but we could not set the error and just have `context canceled: request timed out` as the error message. Thoughts?